enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java

```diff
@@ -979,38 +1006,14 @@ public String toString()
         {
             return format( "{clusterId: %s, message: %s}", clusterId, message() );
         }
-
-        @Override
-        public MemberId from()
-        {
-            return message().from();
-        }
-
-        @Override
-        public Type type()
-        {
-            return message().type();
-        }
-
-        @Override
-        public RaftMessage message()
-        {
-            return message;
-        }
-
-        @Override
-        public <T, E extends Exception> T dispatch( Handler<T,E> visitor ) throws E
-        {
-            return message().dispatch( visitor );
-        }
     }
 
-    class ReceivedInstantAwareMessageImpl implements ReceivedInstantAwareMessage
+    class ReceivedInstantAwareMessageImpl<RM extends RaftMessage> implements ReceivedInstantAwareMessage<RM>
```

Comment:
I wonder if it would be possible and beneficial to flatten the type-hierarchy of raft messages, to just have a single base RaftMessage type with all the different behaviours in it; ReceivedInstant, default implementations of dispatch, ClusterId, and so on.

Reply:
Not all RaftMessage instances have (or need) a ReceivedInstant, or a ClusterId, or either. This way keeps all the complexity in one place.